File: app/src/main/java/com/album/sample/camera/SimpleCameraActivity.java
package com.album.sample.camera;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.ActivityCompat;
import android.support.v4.app.FragmentManager;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.Toast;

import com.album.sample.R;
import com.album.util.FileUtils;
import com.google.android.cameraview.AspectRatio;
import com.google.android.cameraview.CameraView;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.OutputStream;
import java.util.Set;

/**
 * by y on 28/08/2017.
 */

public class SimpleCameraActivity extends AppCompatActivity implements
        ActivityCompat.OnRequestPermissionsResultCallback,
        SimpleAspectRatioFragment.Listener {

    private static final String TAG = "MainActivity";

    private static final String FRAGMENT_DIALOG = "dialog";

    private static final int[] FLASH_OPTIONS = {
            CameraView.FLASH_AUTO,
            CameraView.FLASH_OFF,
            CameraView.FLASH_ON,
    };

    private static final int[] FLASH_ICONS = {
            R.drawable.ic_flash_auto,
            R.drawable.ic_flash_off,
            R.drawable.ic_flash_on,
    };

    private static final int[] FLASH_TITLES = {
            R.string.flash_auto,
            R.string.flash_off,
            R.string.flash_on,
    };

    private int mCurrentFlash;
    private CameraView mCameraView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_simple_camera);
        mCameraView = (CameraView) findViewById(R.id.camera);
        findViewById(R.id.take_picture).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                mCameraView.takePicture();
            }
        });
        Toolbar toolbar = (Toolbar) findViewById(R.id.album_camer_atoolbar);
        setSupportActionBar(toolbar);
        ActionBar actionBar = getSupportActionBar();
        if (actionBar != null) {
            actionBar.setDisplayShowTitleEnabled(false);
        }
        initCamera();
    }

    @Override
    protected void onResume() {
        super.onResume();
        mCameraView.start();
    }

    @Override
    protected void onPause() {
        mCameraView.stop();
        super.onPause();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.simple_camera_menu, menu);
        return true;
    }


    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.aspect_ratio:
                FragmentManager fragmentManager = getSupportFragmentManager();
                if (mCameraView != null
                        && fragmentManager.findFragmentByTag(FRAGMENT_DIALOG) == null) {
                    final Set<AspectRatio> ratios = mCameraView.getSupportedAspectRatios();
                    final AspectRatio currentRatio = mCameraView.getAspectRatio();
                    SimpleAspectRatioFragment.newInstance(ratios, currentRatio)
                            .show(fragmentManager, FRAGMENT_DIALOG);
                }
                return true;
            case R.id.switch_flash:
                if (mCameraView != null) {
                    mCurrentFlash = (mCurrentFlash + 1) % FLASH_OPTIONS.length;
                    item.setTitle(FLASH_TITLES[mCurrentFlash]);
                    item.setIcon(FLASH_ICONS[mCurrentFlash]);
                    mCameraView.setFlash(FLASH_OPTIONS[mCurrentFlash]);
                }
                return true;
            case R.id.switch_camera:
                if (mCameraView != null) {
                    int facing = mCameraView.getFacing();
                    mCameraView.setFacing(facing == CameraView.FACING_FRONT ? CameraView.FACING_BACK : CameraView.FACING_FRONT);
                }
                return true;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onAspectRatioSelected(@NonNull AspectRatio ratio) {
        if (mCameraView != null) {
            Toast.makeText(this, ratio.toString(), Toast.LENGTH_SHORT).show();
            mCameraView.setAspectRatio(ratio);
        }
    }

    private void initCamera() {
        mCameraView.addCallback(new CameraView.Callback() {
            @Override
            public void onCameraOpened(CameraView cameraView) {
                super.onCameraOpened(cameraView);
                Log.d(TAG, "onCameraOpened");
            }

            @Override
            public void onCameraClosed(CameraView cameraView) {
                super.onCameraClosed(cameraView);
                Log.d(TAG, "onCameraClosed");
            }

            @Override
            public void onPictureTaken(final CameraView cameraView, final byte[] data) {
                super.onPictureTaken(cameraView, data);
                Toast.makeText(cameraView.getContext(), R.string.picture_taken, Toast.LENGTH_SHORT).show();
                cameraView.post(new Runnable() {
                    @Override
                    public void run() {
                        File cameraFile = FileUtils.getCameraFile(cameraView.getContext(), null);
                        OutputStream os;
                        try {
                            os = new FileOutputStream(cameraFile);
                            os.write(data);
                            os.close();
                            FileUtils.finishCamera(SimpleCameraActivity.this, cameraFile.getPath());
                        } catch (IOException e) {
                            Log.w(TAG, "Cannot write to " + cameraFile.getPath(), e);
                        }
                    }
                });
            }
        });

    }

    @Override
    public void onBackPressed() {
        super.onBackPressed();
        Toast.makeText(this, "cancel camera", Toast.LENGTH_SHORT).show();
    }


}

File: app/src/main/java/com/album/sample/camera/SimpleCameraActivity.kt
package com.album.sample.camera

import android.os.Bundle
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.Toast
import androidx.core.app.ActivityCompat
import com.album.core.albumPathFile
import com.album.core.finishCamera
import com.album.core.ui.AlbumBaseActivity
import com.album.sample.R
import com.google.android.cameraview.AspectRatio
import com.google.android.cameraview.CameraView
import kotlinx.android.synthetic.main.activity_simple_camera.*
import java.io.FileOutputStream
import java.io.IOException
import java.io.OutputStream

/**
 * by y on 28/08/2017.
 */

class SimpleCameraActivity : AlbumBaseActivity(), ActivityCompat.OnRequestPermissionsResultCallback, SimpleAspectRatioFragment.Listener {
    override fun initView() {
        findViewById<View>(R.id.take_picture).setOnClickListener { camera.takePicture() }
    }

    override fun initCreate(savedInstanceState: Bundle?) {
        setSupportActionBar(toolbar)
        val actionBar = supportActionBar
        actionBar?.setDisplayShowTitleEnabled(false)
        initCamera()
    }

    override val layoutId: Int = R.layout.activity_simple_camera

    private var mCurrentFlash: Int = 0

    override fun onResume() {
        super.onResume()
        camera.start()
    }

    override fun onPause() {
        camera.stop()
        super.onPause()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.simple_camera_menu, menu)
        return true
    }


    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.aspect_ratio -> {
                val fragmentManager = supportFragmentManager
                if (fragmentManager.findFragmentByTag(FRAGMENT_DIALOG) == null) {
                    val ratios = camera.supportedAspectRatios
                    val currentRatio = camera.aspectRatio
                    SimpleAspectRatioFragment.newInstance(ratios, currentRatio!!)
                            .show(fragmentManager, FRAGMENT_DIALOG)
                }
                return true
            }
            R.id.switch_flash -> {
                mCurrentFlash = (mCurrentFlash + 1) % FLASH_OPTIONS.size
                item.setTitle(FLASH_TITLES[mCurrentFlash])
                item.setIcon(FLASH_ICONS[mCurrentFlash])
                camera.flash = FLASH_OPTIONS[mCurrentFlash]
                return true
            }
            R.id.switch_camera -> {
                val facing = camera.facing
                camera.facing = if (facing == CameraView.FACING_FRONT) CameraView.FACING_BACK else CameraView.FACING_FRONT
                return true
            }
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onAspectRatioSelected(ratio: AspectRatio) {
        Toast.makeText(this, ratio.toString(), Toast.LENGTH_SHORT).show()
        camera.setAspectRatio(ratio)
    }

    private fun initCamera() {
        camera.addCallback(object : CameraView.Callback() {
            override fun onCameraOpened(cameraView: CameraView) {
                super.onCameraOpened(cameraView)
                Log.d(TAG, "onCameraOpened")
            }

            override fun onCameraClosed(cameraView: CameraView) {
                super.onCameraClosed(cameraView)
                Log.d(TAG, "onCameraClosed")
            }

            override fun onPictureTaken(cameraView: CameraView, data: ByteArray) {
                super.onPictureTaken(cameraView, data)
                Toast.makeText(cameraView.context, R.string.picture_taken, Toast.LENGTH_SHORT).show()
                cameraView.post {
                    val cameraFile = albumPathFile(null, System.currentTimeMillis().toString(), "jpg")
                    val os: OutputStream
                    try {
                        os = FileOutputStream(cameraFile)
                        os.write(data)
                        os.close()
                        finishCamera(cameraFile)
                    } catch (e: IOException) {
                        Log.w(TAG, "Cannot write to $cameraFile", e)
                    }
                }
            }
        })

    }

    override fun onBackPressed() {
        super.onBackPressed()
        Toast.makeText(this, "cancel camera", Toast.LENGTH_SHORT).show()
    }

    companion object {

        private const val TAG = "MainActivity"

        private const val FRAGMENT_DIALOG = "dialog"

        private val FLASH_OPTIONS = intArrayOf(CameraView.FLASH_AUTO, CameraView.FLASH_OFF, CameraView.FLASH_ON)

        private val FLASH_ICONS = intArrayOf(R.drawable.ic_flash_auto, R.drawable.ic_flash_off, R.drawable.ic_flash_on)

        private val FLASH_TITLES = intArrayOf(R.string.flash_auto, R.string.flash_off, R.string.flash_on)
    }


}
